File: packages/swift-client/Sources/PyricDebugUI/RulesDenialReport.swift
import Foundation
import PyricFirestore

/// Structured representation of a Security Rules evaluation rejection (CEL denial).
public struct RulesDenialReport: Identifiable, Sendable, Equatable {
    public let id: UUID
    public let timestamp: Date
    public let file: String
    public let line: Int?
    public let col: Int?
    public let citation: String
    public let expression: String?
    public let reasons: [String]
    public let authUid: String?
    public let authTenant: String?
    public let authClaims: [String: AnySendable]?
    public let requestMethod: String?
    public let requestPath: String?
    public let proposedData: [String: AnySendable]?
    public let existingData: [String: AnySendable]?
    public let failedFields: [String]
    public let query: AnySendable?
    public let errorMessage: String

    public init(
        id: UUID = UUID(),
        timestamp: Date = Date(),
        file: String = "firestore.rules",
        line: Int? = nil,
        col: Int? = nil,
        citation: String? = nil,
        expression: String? = nil,
        reasons: [String] = [],
        authUid: String? = nil,
        authTenant: String? = nil,
        authClaims: [String: AnySendable]? = nil,
        requestMethod: String? = nil,
        requestPath: String? = nil,
        proposedData: [String: AnySendable]? = nil,
        existingData: [String: AnySendable]? = nil,
        failedFields: [String] = [],
        query: AnySendable? = nil,
        errorMessage: String = "Missing or insufficient permissions."
    ) {
        self.id = id
        self.timestamp = timestamp
        self.file = file
        self.line = line
        self.col = col
        if let citation, !citation.isEmpty {
            self.citation = citation
        } else if let line {
            if let col {
                self.citation = "\(file):\(line):\(col)"
            } else {
                self.citation = "\(file):\(line)"
            }
        } else {
            self.citation = file
        }
        self.expression = expression
        self.reasons = reasons
        self.authUid = authUid
        self.authTenant = authTenant
        self.authClaims = authClaims
        self.requestMethod = requestMethod
        self.requestPath = requestPath
        self.proposedData = proposedData
        self.existingData = existingData
        self.failedFields = failedFields
        self.query = query
        self.errorMessage = errorMessage
    }

    /// Constructs a `RulesDenialReport` from raw wire `denialContext` payload and error message.
    public static func from(denialContext: AnySendable, message: String = "Missing or insufficient permissions.") -> RulesDenialReport {
        let ruleObj = denialContext["rule"]
        let rawFile = ruleObj?["file"]?.stringValue ?? "firestore.rules"
        let line = ruleObj?["line"]?.intValue.map { Int($0) }
        let col = (ruleObj?["col"]?.intValue ?? ruleObj?["column"]?.intValue).map { Int($0) }
        let citation = ruleObj?["citation"]?.stringValue
        let expression = ruleObj?["expression"]?.stringValue

        var reasons: [String] = []
        if let reasonsArray = denialContext["reasons"]?.arrayValue {
            reasons = reasonsArray.compactMap { $0.stringValue }
        }
        if reasons.isEmpty {
            reasons = [message]
        }

        let authObj = denialContext["auth"]
        let authUid = authObj?["uid"]?.stringValue
        let authClaims = authObj?["token"]?.dictionaryValue
        let authTenant = authObj?["token"]?["firebase"]?["tenant"]?.stringValue

        let reqObj = denialContext["request"]
        let reqMethod = reqObj?["method"]?.stringValue
        let reqPath = reqObj?["path"]?.stringValue
        let proposedData = reqObj?["resourceData"]?.dictionaryValue

        let resObj = denialContext["resource"]
        let existingData = resObj?["data"]?.dictionaryValue

        var failedFields: [String] = []
        if let fieldsArray = denialContext["failedFields"]?.arrayValue {
            failedFields = fieldsArray.compactMap { $0.stringValue }
        }

        let query = denialContext["query"]

        return RulesDenialReport(
            file: rawFile,
            line: line,
            col: col,
            citation: citation,
            expression: expression,
            reasons: reasons,
            authUid: authUid,
            authTenant: authTenant,
            authClaims: authClaims,
            requestMethod: reqMethod,
            requestPath: reqPath,
            proposedData: proposedData,
            existingData: existingData,
            failedFields: failedFields,
            query: query,
            errorMessage: message
        )
    }

    /// Extracts a `RulesDenialReport` from a `PyricBridgeError` if it carries a `denialContext`.
    public static func from(error: PyricBridgeError) -> RulesDenialReport? {
        guard let denialContext = error.denialContext else { return nil }
        return from(denialContext: denialContext, message: error.message)
    }
}

File: packages/flutter-client/lib/src/debug/rules_denial_report.dart
import 'package:collection/collection.dart';
import '../transport/exceptions.dart';

String? _asString(dynamic v) => v == null ? null : (v is String ? v : v.toString());

int? _asInt(dynamic v) {
  if (v == null) return null;
  if (v is int) return v;
  if (v is num) return v.toInt();
  return int.tryParse(v.toString());
}

Map<String, dynamic>? _asStringMap(dynamic v) {
  if (v is! Map) return null;
  final result = <String, dynamic>{};
  for (final entry in v.entries) {
    if (entry.key != null) {
      result[entry.key.toString()] = entry.value;
    }
  }
  return result;
}

List<String> _asStringList(dynamic v) {
  if (v is! List) return [];
  final result = <String>[];
  for (final item in v) {
    if (item != null) {
      result.add(item.toString());
    }
  }
  return result;
}

/// Structured representation of a Security Rules evaluation failure (CEL rejection).
class RulesDenialReport {
  final String file;
  final int? line;
  final int? col;
  final String citation;
  final String? expression;
  final List<String> reasons;
  final String? authUid;
  final String? authTenant;
  final Map<String, dynamic>? authClaims;
  final String? requestMethod;
  final String? requestPath;
  final Map<String, dynamic>? proposedData;
  final Map<String, dynamic>? existingData;
  final List<String> failedFields;
  final Map<String, dynamic>? query;
  final String errorMessage;
  final DateTime timestamp;

  RulesDenialReport({
    this.file = 'firestore.rules',
    this.line,
    this.col,
    String? citation,
    this.expression,
    this.reasons = const [],
    this.authUid,
    this.authTenant,
    this.authClaims,
    this.requestMethod,
    this.requestPath,
    this.proposedData,
    this.existingData,
    this.failedFields = const [],
    this.query,
    this.errorMessage = 'Missing or insufficient permissions.',
    DateTime? timestamp,
  })  : citation = citation != null && citation.isNotEmpty
            ? citation
            : line != null
                ? col != null
                    ? '$file:$line:$col'
                    : '$file:$line'
                : file,
        timestamp = timestamp ?? DateTime.now();

  /// Constructs a `RulesDenialReport` from a `denialContext` map and error message.
  factory RulesDenialReport.fromMap(
    Map<dynamic, dynamic> map, {
    String errorMessage = 'Missing or insufficient permissions.',
  }) {
    final ruleObj = _asStringMap(map['rule']);
    final file = _asString(ruleObj?['file']) ?? 'firestore.rules';
    final line = _asInt(ruleObj?['line']);
    final col = _asInt(ruleObj?['col']) ?? _asInt(ruleObj?['column']);
    final citation = _asString(ruleObj?['citation']);
    final expression = _asString(ruleObj?['expression']);

    final reasons = _asStringList(map['reasons']);
    if (reasons.isEmpty) {
      reasons.add(errorMessage);
    }

    final authObj = _asStringMap(map['auth']);
    final authUid = _asString(authObj?['uid']);
    final authClaims = _asStringMap(authObj?['token']);
    final firebaseMap = _asStringMap(authClaims?['firebase']);
    final authTenant = _asString(firebaseMap?['tenant']) ?? _asString(authObj?['tenant']);

    final reqObj = _asStringMap(map['request']);
    final requestMethod = _asString(reqObj?['method']);
    final requestPath = _asString(reqObj?['path']);
    final proposedData = _asStringMap(reqObj?['resourceData']);

    final resObj = _asStringMap(map['resource']);
    final existingData = _asStringMap(resObj?['data']);

    final failedFields = _asStringList(map['failedFields']);

    final query = _asStringMap(map['query']);

    return RulesDenialReport(
      file: file,
      line: line,
      col: col,
      citation: citation,
      expression: expression,
      reasons: reasons,
      authUid: authUid,
      authTenant: authTenant,
      authClaims: authClaims,
      requestMethod: requestMethod,
      requestPath: requestPath,
      proposedData: proposedData,
      existingData: existingData,
      failedFields: failedFields,
      query: query,
      errorMessage: errorMessage,
    );
  }

  /// Extracts a `RulesDenialReport` from a `PyricBridgeException` if it contains `denialContext`.
  static RulesDenialReport? fromBridgeException(PyricBridgeException ex) {
    if (ex.denialContext is! Map) return null;
    return RulesDenialReport.fromMap(
      ex.denialContext as Map,
      errorMessage: ex.message,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is RulesDenialReport &&
        other.citation == citation &&
        other.expression == expression &&
        other.authUid == authUid &&
        const ListEquality().equals(other.reasons, reasons);
  }

  @override
  int get hashCode => Object.hash(
        citation,
        expression,
        authUid,
        const ListEquality().hash(reasons),
      );
}
